### PR TITLE
Relocate Kindling

### PR DESCRIPTION
**Pull request by airidaceae**
Relocating kindling into the main repository.
to-do:
- [x] create init system

follow up:
- [ ] archive https://github.com/hearth-rs/kindling
- [x] issues to open:
  - [x] CI tests for kindling
  - [x] move helper code to common crate
  - [x] general behavior crates
  - [x] readme

### CODE REVIEW

File: kindling/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[workspace]
+resolver = "2"
+members = ["panic-test", "init"]
```
Comment (marceline-cramer) [inline kindling/Cargo.toml:3]: While we're refactoring, could we move all the services into a `services` subdirectory and include into the workspace members list with `services/*`?

File: kindling/build/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2023 the Hearth contributors.
```
Comment (marceline-cramer) [inline kindling/build/src/main.rs:1]: Do you wanna add your copyright here?

File: kindling/build/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                break;
+            }
+        }
+
+        if !is_lib {
+            eprintln!("{:?} is not a lib; skipping", package.name);
+            continue;
+        }
+
+        build_service(&root_path, &package);
```
Comment (marceline-cramer) [inline kindling/build/src/main.rs:56]: clippy

File: kindling/build/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    let _ = std::fs::create_dir_all(path);
+    true
+}
+
+fn get_cargo() -> String {
+    std::env::var("CARGO").expect("CARGO env var isn't set")
+}
+
+fn get_kindling_dir() -> String {
+    std::env::var("CARGO_WORKSPACE_DIR").expect("CARGO_WORKSPACE_DIR env var isn't set")
```
Comment (marceline-cramer) [inline kindling/build/src/main.rs:73]: I love this. Really clever.

File: kindling/build/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        + "kindling/"
+}
+
+fn build_service(root_path: &Path, package: &Package) {
```
Comment (marceline-cramer) [inline kindling/build/src/main.rs:77]: Make sure to open up a follow-up issue to use a common kindling schema crate to write the service metadata to the root.

File: kindling/build/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                            continue;
+                        }
+
+                        eprintln!("copying {:?} to {:?}", file, path);
+                        std::fs::copy(file, path).unwrap();
+                    }
+                }
+            }
+            _ => {}
+        }
```
Comment (marceline-cramer) [inline kindling/build/src/main.rs:141]: clippy

File: kindling/init/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "init"
```
Comment (marceline-cramer) [inline kindling/init/Cargo.toml:2]: kindling-init

File: kindling/init/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+// details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with Hearth. If not, see <https://www.gnu.org/licenses/>.
+
+// Currently just a stub for testing purposes
+#[no_mangle]
+pub extern "C" fn run() {
+    panic!("A rather sad init system")
```
Comment (marceline-cramer) [inline kindling/init/src/lib.rs:22]: can relate

File: kindling/panic-test/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "panic-test"
```
Comment (marceline-cramer) [inline kindling/panic-test/Cargo.toml:2]: kindling-panic-test

**Review by marceline-cramer — CHANGES_REQUESTED** [review]
Looking great.

File: kindling/init/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2023 the Hearth contributors.
```
Comment (marceline-cramer) [inline kindling/init/src/lib.rs:1]: Copyright?

File: kindling/panic-test/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2023 the Hearth contributors.
```
Comment (marceline-cramer) [inline kindling/panic-test/src/lib.rs:1]: Copyright?

**Review by marceline-cramer — CHANGES_REQUESTED** [review]

**Comment by marceline-cramer** [comment]
Also add a follow-up for adding kindling to CI.

File: kindling/build/src/main.rs (hunk 5 above)
Comment (airidaceae) [reply kindling/build/src/main.rs:77]: added to the follow up list

**Comment by airidaceae** [comment]
moved the wasm cargo target to each individual service because if its in the root it breaks the kindling build. There's no way to reset the build target yet and I'm not holding my breath. https://github.com/rust-lang/cargo/issues/8687

**Comment by marceline-cramer** [comment]
That probably explains why the wasm target wasn't set in the upstream repo.

File: kindling/init/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            wasm::WasmSpawnInfo {
+                lump,
+                entrypoint: None,
+            },
+            &[REGISTRY.as_ref()],
+        );
+    }
+}
+
+//TODO break these file system operations into
```
Comment (marceline-cramer) [inline kindling/init/src/lib.rs:57]: into what?

**Review by marceline-cramer — APPROVED** [review]

File: kindling/init/src/lib.rs (hunk 12 above)
Comment (airidaceae) [reply kindling/init/src/lib.rs:57]: in my defense. i got sleepy

**Pull request merged**